TechnicalDocuments/Proposals/IdentifiersForUIAutomation.md

```diff
@@ -0,0 +1,203 @@
+
+Components are free to set accessibility identifiers in whatever way they see fit, but they *must* append the given prefix to any accessibility identifier it set.
+
+Bento shall generates a prefix based on the section ID and the item ID. All variables are suffixed by `/`. So for example, the prefix for item `card` of section `activity` should be `activity/card/`.
+
+### Working with Atomic Components and Design Library
+
+Atomic components present a challenge, since composition is allowed and thus multiple components may present at the same ID path.
+
+The proposal suggests that all atomic components allow a custom _automation name_ to be specified, and implements `ViewLifecycleAware.updateAccessibilityIdentifiers(usingPrefix:)`. Composite components, that have multiple opaque child components, should propagate this call to its children if appropriate.
+
+The Design Library shall introduce a new umbrella namespace `DesignLibrary.AutomationName` for all defined automation names:
+
+```swift
+extension DesignLibrary {
+    public enum AutomationName {}
+}
+```
+
+For example, given the conformance of these two atomic components:
+
+```swift
+struct UI.Label {
+    // ... other properties
+    let automationName: String
+
+    init(/* ..., */automationName: String = "label") {
+        self.automationName = automationName
+    }
+}
+
+struct UI.Button {
+    // ... other properties
+    let automationName: String
+
+    init(/* ..., */automationName: String = "button") {
+        self.automationName = automationName
+    }
+}
+
+extension UI.Label: ViewLifecycleAware {
+    func updateAccessibilityIdentifiers(usingPrefix prefix: String) {
+        return prefix + automationName
+    }
+}
+
+extension UI.Button: ViewLifecycleAware {
+    func updateAccessibilityIdentifiers(usingPrefix prefix: String) {
+        return prefix + automationName
+    }
+}
+```
+
+and this particular composite component called `prettyCard`:
+
+```swift
+extension DesignLibrary.ComponentsLibrary {
+    public func prettyCard() -> AnyRenderable {
+        return [
+            UI.Label(/* ... */, automationName: String(AutomationName.prettyCardNotice)),
+            UI.Button(/* ... */, automationName: String(AutomationName.prettyCardSubmit))
+        ].stack(axis: .vertical)
+    }
+}
+
+extension DesignLibrary.AutomationName {
+    public enum PrettyCard: String {
+        case notice
+        case submit
+    }
+}
+```
+
+When we render `prettyCard`:
+
+```
+render(
+    Box(
+        sections: [
+            Section(id: .activity, items: [
+                Node(id: .card, component:
+                    prettyCard()
+                )
+            ])
+        ]
+    )
+)
+```
+
+It should result in two accessibility identifiers:
+
+* `activity/card/notice` applied to the label; and
```

Comment:
I could be wrong but by label you mean "accessibilityLabel", we need to be careful here as the label is visible to the end user and used by VoiceOver, or did I misunderstand?

Reply:
I mean the `UILabel`.